src/ranged.cpp

```diff
@@ -619,11 +570,17 @@ int player::fire_gun( const tripoint &target, int shots, item& gun )
         }
     }
 
-    practice( skill_gun, train_skill ? 15 : 0 );
-
     // Use different amounts of time depending on the type of gun and our skill
     moves -= time_to_fire( *this, *gun.type );
 
+    practice( skill_gun, xp * ( get_skill_level( skill_gun ) + 1 ) );
+    if( !xp && one_in( 10 ) ) {
+        add_msg_if_player( m_info, _( "You'll need to aim at more distant targets to further improve your marksmanship." ) );
```

Comment:
This prints even if the player aims at the edge of the bubble and fails to hit the targets.